File: exonum-java-binding-core/rust/src/storage/memorydb.rs
// Copyright 2018 The Exonum Team
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use exonum::storage::{Database, MemoryDB};
use jni::JNIEnv;
use jni::objects::{JClass, JObject};

use std::panic;

use storage::db::{View, ViewRef};
use utils::{self, Handle};

/// Returns pointer to created `MemoryDB` object.
#[no_mangle]
pub extern "system" fn Java_com_exonum_binding_storage_database_MemoryDb_nativeCreate(
    env: JNIEnv,
    _: JClass,
) -> Handle {
    let res = panic::catch_unwind(|| Ok(utils::to_handle(MemoryDB::new())));
    utils::unwrap_exc_or_default(&env, res)
}

/// Destroys underlying `MemoryDB` object and frees memory.
#[no_mangle]
pub extern "system" fn Java_com_exonum_binding_storage_database_MemoryDb_nativeFree(
    env: JNIEnv,
    _: JClass,
    db_handle: Handle,
) {
    utils::drop_handle::<MemoryDB>(&env, db_handle);
}

/// Returns pointer to created `Snapshot` object.
#[no_mangle]
pub extern "system" fn Java_com_exonum_binding_storage_database_MemoryDb_nativeCreateSnapshot(
    env: JNIEnv,
    _: JObject,
    db_handle: Handle,
) -> Handle {
    let res = panic::catch_unwind(|| {
        let db = utils::cast_handle::<MemoryDB>(db_handle);
        Ok(utils::to_handle(View::from_owned_snapshot(db.snapshot())))
    });
    utils::unwrap_exc_or_default(&env, res)
}

/// Returns pointer to created `Fork` object.
#[no_mangle]
pub extern "system" fn Java_com_exonum_binding_storage_database_MemoryDb_nativeCreateFork(
    env: JNIEnv,
    _: JObject,
    db_handle: Handle,
) -> Handle {
    let res = panic::catch_unwind(|| {
        let db = utils::cast_handle::<MemoryDB>(db_handle);
        Ok(utils::to_handle(View::from_owned_fork(db.fork())))
    });
    utils::unwrap_exc_or_default(&env, res)
}

/// Merges the given fork into the database.
#[no_mangle]
pub extern "system" fn Java_com_exonum_binding_storage_database_MemoryDb_nativeMerge(
    env: JNIEnv,
    _: JObject,
    db_handle: Handle,
    view_handle: Handle,
) {
    let res = panic::catch_unwind(|| {
        let db = utils::cast_handle::<MemoryDB>(db_handle);
        let fork = match *utils::cast_handle::<View>(view_handle).get() {
            ViewRef::Snapshot(_) => panic!("Attempt to merge snapshot instead of fork."),
            ViewRef::Fork(ref fork) => fork,
        };
        db.merge(fork.patch().clone()).expect(
            "Unable to merge fork",
        );
        Ok(())
    });
    utils::unwrap_exc_or_default(&env, res)
}

File: exonum-java-binding-core/src/main/java/com/exonum/binding/storage/database/MemoryDb.java
/* 
 * Copyright 2018 The Exonum Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.exonum.binding.storage.database;

import static com.exonum.binding.proxy.NativeHandle.INVALID_NATIVE_HANDLE;

import com.exonum.binding.proxy.AbstractCloseableNativeProxy;
import com.exonum.binding.proxy.Cleaner;
import com.google.common.annotations.VisibleForTesting;

/**
 * An in-memory database for testing purposes. It can create both read-only snapshots
 * and read-write forks. The changes made to database forks can be applied to the database state.
 */
@VisibleForTesting
public final class MemoryDb extends AbstractCloseableNativeProxy implements Database {

  /**
   * Creates a new empty MemoryDb.
   */
  public static MemoryDb newInstance() {
    long nativeHandle = INVALID_NATIVE_HANDLE;
    try {
      nativeHandle = nativeCreate();
      return new MemoryDb(nativeHandle);
    } catch (Throwable t) {
      if (nativeHandle != INVALID_NATIVE_HANDLE) {
        nativeFree(nativeHandle);
      }
      throw t;
    }
  }

  @VisibleForTesting  // Used in native resource manager tests, must not be exported.
  MemoryDb(long nativeHandle) {
    super(nativeHandle, true);
  }

  @Override
  public Snapshot createSnapshot(Cleaner cleaner) {
    long snapshotHandle = nativeCreateSnapshot(getNativeHandle());
    return Snapshot.newInstance(snapshotHandle, cleaner);
  }

  @Override
  public Fork createFork(Cleaner cleaner) {
    long forkHandle = nativeCreateFork(getNativeHandle());
    return Fork.newInstance(forkHandle, cleaner);
  }

  /**
   * Applies the changes from the given fork to the database state.
   *
   * @param fork a fork to get changes from
   */
  public void merge(Fork fork) {
    nativeMerge(getNativeHandle(), fork.getViewNativeHandle());
  }

  @Override
  protected void disposeInternal() {
    nativeFree(getNativeHandle());
  }

  private static native long nativeCreate();

  private native long nativeCreateSnapshot(long dbNativeHandle);

  private native long nativeCreateFork(long dbNativeHandle);

  private native void nativeMerge(long dbNativeHandle, long forkNativeHandle);

  private static native void nativeFree(long dbNativeHandle);
}
